mathics/builtin/arithmetic.py

```diff
@@ -1861,6 +1861,60 @@ class Factorial(PostfixOperator, _MPMathFunction):
     mpmath_name = "factorial"
 
 
+class Factorial2(PostfixOperator, _MPMathFunction):
+    """
+    <dl>
+      <dt>'Factorial2[$n$]'
+      <dt>'$n$!!'
+      <dd>computes the double factorial of $n$.
+    </dl>
+
+    >> 5!!
+     = 15
+
+    >> Factorial2[-3]
+     = -1
+
+    'Factorial2' also accept irrational numbers:
+    >> N[Pi!!, 6]
+     = 3.35237
+
+    'Factorial2' also accept complex numbers:
+    >> N[I!!, 6]
+     = 2.71713 + 0.279527 I
+
+    'Factorial2' is not defined for even negative numbers.
+    #> (-4)!!
+     : Factorial2 is not defined for even negative numbers.
+     = -4!!
+    """
+
+    operator = "!!"
+    precedence = 610
+    mpmath_name = "fac2"
+    sympy_name = "factorial2"
+
+    messages = {
+        "ndf": "Factorial2 is not defined for even negative numbers.",
+        "unknownp": "'`1`' not in ('sympy', 'numpy', 'mpmath')",
+    }
+
+    def apply(self, number, evaluation, preference="sympy"):
+        "Factorial2[number_?NumberQ]"
+
+        if not (preference in ("sympy", "numpy", "mpmath")):
+            return evaluation.message("Factorial2", "unknownp", preference)
+
+        if isinstance(number, Integer) and (preference != "mpmath"):
+            if number.to_python() < 0 and number.to_python() % 2 == 0:
```

Comment:
Something to think about.  Another possibility is to pass to sympy and report back errors it gives.

That way if/when sympy handles negative even numbers, we will pick that up and not have to code here as we would do by being proactive about this. 

Not a big deal, but something to keep in mind.

Reply:
@rocky the error is not so user-friendly. It says it's in a pole of the gamma function.

Reply:
Ok but you could still trap the error and _then_ check for negativity. And if it's not that, use whatever sympy has to say.

Again that way, if sympy adapts, you do. And having code in place to trap for the error might be used if there are _other_ kinds of messages you want to redo.

Reply:
@rock I don't understand why do it.

My current approach is the following:
```py
def apply(self, number, evaluation):
    "Factorial2[number_?NumberQ]"

    if number.to_python() < 0 and number.to_python() % 2 == 0:
        # even negative number
        return evaluation.message("Factorial2", "ndf")

    if isinstance(number, Integer):
        return from_sympy(getattr(sympy, self.sympy_name)(number.to_sympy()))
    else:
        return from_mpmath(getattr(mpmath, self.mpmath_name)(number.to_mpmath()))
```

For either SymPy or MPMath the double factorial of a negative odd leads to an error.

Reply:
In https://en.wikipedia.org/wiki/Double_factorial negative even is defined. It is just not _currently_ implemented. 

Do you know it will _never_  be implemented in _either_ of these? 

Again, I'd like to think of this code as not offering opinions so much as just acting as a pass through to an unbounded open set of other libraries. We let the other libraries be the experts on the thing they do. (But there is more than one so, we don't really have to rely on that totally). 

The attitude that this code is going to implement all of graphics routines, or its own documentation system has just got us into trouble time and time again.

Reply:
@rocky I'm saying that if we write this and then one of them implements it we would remove the `ndf` error, so we're going to change the code anyway.

Reply:
But users get the benefit right away. Removing what will be dead code has little impact here. 

But whatever - beggars can't be choosers. I suppose this is fine and it can be redone in the future if desired.

Reply:
@rocky I'm afraid that first trying and then checking for the negativity would be too complex (and it'd be done twice), and the difficulty of removing the `ndf` is going to be smaller than removing the check.

Anyway, you're working in Mathics for much more time than me, so I guess your decision is going to be the best. What do you say?

Reply:
> @rocky I'm afraid that first trying and then checking for the negativity would be too complex (and it'd be done twice), 

It's exactly the same code test: `if number.to_python() < 0 and number.to_python() % 2 == 0:`

It would just be moved whereever failure is detected. 

When there is an error very little is being done so I don't see it as a big deal. And the second test is there just to provide a more sensible error message. 

> and the difficulty of removing the `ndf` is going to be smaller than removing the check.
> 
> Anyway, you're working in Mathics for much more time than me, so I guess your decision is going to be the best. What do you say?

Since seems to be a big deal and it should be a big deal either way, just do it however you want. We can adjust later if it becomes important, which I don't think it will.

Reply:
But one thing please keep in mind because it has come up time and time again. Don't try to second guess the libraries. If you want to improve them, go to the respective repositories and suggest patches and PRs. And in that process you may come to learn more about why things are the way they are and subtleties that you may have missed. 

I go about changing or "fixing" the sympy/numpy/mpmath library code very humbly.

Reply:
@rocky the try didn't increase the complexity of the code too much (to say the truth it didn't at all, the pseudocode in my mind was much more complex).

About fixing MPMath, this sounds like a great idea! I'll give it a try.

Reply:
Ok - I'll look at this when I get a chance.

Reply:
> Ok - I'll look at this when I get a chance.

https://github.com/mathics/Mathics/pull/1459 is what i have. It is a bit messier than I had hoped because there are bugs outside of just the Factorial2 part.

But this brings up another point when you just jump off into an areas that you've never looked at before. Be prepared to find bugs and uncover a new can of worms. 

As you are aware, we're trying to work through all of these. But the misfeature are still vast.